File: EnergizeMarkov/src/markov_chain.rs
use std::path::Path;
use std::fs;
use std::fs::OpenOptions;
use std::io::BufWriter;
use std::io::BufReader;
use std::io::Write;
use std::io::BufRead;
use rand;
use regex::Regex;
use rand::Rng;
use std::fs::File;

pub struct MarkovChain
{
    path: String,
    max_depth: u32,
    extension: String,
}

impl MarkovChain
{
    pub fn new(path: String,max_depth: u32,extension: String) -> Self
    {
        MarkovChain
        {
            path,
            max_depth,
            extension,
        }
    }

    fn append_all_text(&self,path: String,mut text: String) -> bool
    {
        text.push('\n');
        let fileres = OpenOptions::new()
            .write(true)
            .create(true)
            .append(true)
            .open(path);

        if let Err(_) = fileres
        {
            return false;
        }

        let file = fileres.unwrap();
        let mut writer = BufWriter::new(&file);
        match writer.write(text.as_bytes())
        {
            Ok(_) => true,
            Err(_) => false
        }
    }

    fn read_all_lines(&self,path: String) -> Vec<String>
    {
        let fileres = File::open(path);
        if let Err(_) = fileres
        {
            return vec![];
        }

        let file = fileres.unwrap();
        let reader = BufReader::new(&file);
        let contents = reader.lines().map(|r| r.unwrap());
        
        contents.collect::<Vec<String>>()
    }

    fn split<'a>(&self,sentence: &'a str) -> Vec<&'a str>
    {
        let re = Regex::new(r"\s|\.|,|!|\?|;|_").unwrap();
        re.split(sentence).collect::<Vec<&str>>()
    }

    fn split_noref(&self,sentence: String) -> Vec<String>
    {
        let re = Regex::new(r"\s|\.|,|!|\?|;|_").unwrap();
        let mut parts: Vec<String> = vec![];
        for x in re.split(&sentence)
        {
            parts.push(String::from(x));
        }

        parts
    }

    pub fn learn(&self,mut sentence: String) -> bool
    {
        if sentence == ""
        {
            return false;
        }

        if !Path::new(&self.path).exists()
        {
            if let Err(_) = fs::create_dir(&self.path)
            {
                return false;
            }
        }

        sentence = sentence.trim().to_lowercase();
        let re = Regex::new(r"[a-z]+://\S+").unwrap();
        sentence = re.replace_all(&sentence,"LINK-REMOVED").to_string();
        sentence = sentence.replace("\\","/").replace("/"," ");
        
        let words: Vec<&str> = self.split(&sentence);
        let mut i: i32 = 0;
        let mut worditer = words.iter().peekable();

        while let Some(word) = worditer.next()
        {
            let cur = word.trim();
            let mut next = match worditer.peek()
            {
                Some(x) => x.to_string(),
                None => String::from("END_SEQUENCE")
            };
            next.push('\n');

            let mut path = format!("{}{}{}",self.path,cur,self.extension);
            let mut success = self.append_all_text(path,next.clone());
            if !success { return false; }

            let mut left = format!("{}{}",cur,self.extension);
            for level in 1..self.max_depth
            {
                let depth = i - level as i32;
                if depth >= 0
                {
                    left = format!("{}_{}",words[level as usize].trim(),left);
                    path = format!("{}{}",self.path,left);

                    success = self.append_all_text(path,next.clone());
                    if !success { return false; }
                }
            }
            
            i += 1;
        }

        true
    }

    pub fn generate(&self,wordcount: u32) -> String
    {
        let mut rng = rand::thread_rng();
        let mut files = fs::read_dir(&self.path)
            .unwrap()
            .collect::<Vec<_>>();
        
        match files.is_empty()
        {
            true => 
            {
                let max = files.len() - 1;
                let index = rng.gen_range::<usize>(0,max);
                let entry = files.swap_remove(index).unwrap();
                let word = entry.file_name()
                    .into_string()
                    .unwrap();

                format!("{} {}",word.clone(),self.generate_from(word,wordcount))
            },
            false => String::new()
        }
    }

    pub fn generate_from(&self,firstpart: String,wordcount: u32) -> String
    {
        let mut rng = rand::thread_rng();
        let mut res = String::new();
        let mut curinput = self.split_noref(firstpart);

        for _ in 0..wordcount
        {
            let path = format!("{}{}{}",self.path,curinput.join("_"),self.extension);
            if !Path::new(&path).exists()
            {
                let lines = self.read_all_lines(path);
                let max = lines.len() - 1;
                let index = rng.gen_range::<usize>(0,max);
                let next = &lines[index];

                if next == "END_SEQUENCE" { break; }

                res = format!("{} {}",res,next);
                curinput.push(next.clone());
                if curinput.len() > self.max_depth as usize
                {
                    curinput.remove(0);
                }
            }
            else
            {   
                break;
            }
        }

        res
    }
}

File: Energize/Services/Markov/MarkovChain.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text.RegularExpressions;

namespace Energize.Services.Markov
{
    public class MarkovChain
    {
        private static string _Path = "External/Markov/";
        private static int _MaxDepth = 2;
        private static string _Extension = ".markov";
        private static char[] _Separators = { ' ', '.', ',', '!', '?', ';', '_' };

        public void Learn(string sentence)
        {
            if(sentence == null) return;

            if (!Directory.Exists(_Path))
            {
                Directory.CreateDirectory(_Path);
            }

            sentence = sentence.Trim().ToLower();
            sentence = Regex.Replace(sentence,@"https?://\S*","LINK-REMOVED ");
            sentence = sentence.Replace("\\","/").Replace("/"," ");

            string[] words = sentence.Split(_Separators);

            for (int i = 0; i < words.Length - 1; i++)
            {
                string word = words[i].ToLower().Trim();
                string next = i + 1 > words.Length ? "END_SEQUENCE" : words[i + 1].ToLower().Trim();

                string path = _Path + word + _Extension;
                File.AppendAllText(path, next + "\n");

                string left = word + _Extension;
                for(int depth = 1; depth < _MaxDepth; depth++)
                {
                    if(i - depth >= 0)
                    {
                        left = words[i - depth].Trim() + "_" + left;
                        path = _Path + left;

                        File.AppendAllText(path, next + "\n");
                    }
                }
            }
        }

        public string Generate(int? max=null)
        {
            max = max ?? int.MaxValue;
            Random rand = new Random();
            string[] files = Directory.GetFiles(_Path);

            string word = files[rand.Next(0, files.Length - 1)];
            string[] dirs = word.Split('/');
            word = dirs[dirs.Length - 1];
            word = word.Remove(word.Length - _Extension.Length);

            //Console.WriteLine(word);
            string result = word + ' ' + Generate(word, max.Value);
            result = result.Replace('_',' ');

            return result;
        }

        public string Generate(string firstpart, int wordcount)
        {
            Random rand = new Random();
            string result = "";
            List<string> currentinput = new List<string>(firstpart.Split(_Separators));

            for(int i = 0; i < wordcount; i++)
            {
                string path = _Path + string.Join('_',currentinput) + _Extension;
                if(File.Exists(path))
                {
                    string[] nexts = File.ReadAllLines(path);
                    string next = nexts[rand.Next(0,nexts.Length)].Trim();

                    if(next == "END_SEQUENCE") break;

                    result += " " + next;

                    currentinput.Add(next);
                    if(currentinput.Count > _MaxDepth)
                    {
                        currentinput.RemoveAt(0);
                    }
                }
                else
                {
                    break;
                }
            }

            return result;
        }
    }
}
